File: src/test_programs/test_ipd.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "openmmpol.h"

int countLines(char *fin){
    FILE *fp = fopen(fin, "r");
    
    char c;
    int lines = 1;

    if(fp == NULL) return 0;

    do{
        c = fgetc(fp);
        if(c == '\n') lines++;
    }while(c != EOF);

    fclose(fp);
  
    return lines - 1;
}

double **read_ef(char *fin){
    double **ef;
    
    int pol_atoms = countLines(fin);

    ef = (double **) malloc(sizeof(double *) * 3 * pol_atoms);

    FILE *fp = fopen(fin, "r");
    for(int i =0; i < pol_atoms; i++){
        ef[i] = (double *) malloc(sizeof(double) * 3);
        fscanf(fp, "%lf %lf %lf", &(ef[i][0]), &(ef[i][1]),  &(ef[i][2]));
    }
    fclose(fp);

    return ef;
}

int main(int argc, char **argv){
    if(argc != 4 && argc != 5){
        printf("Syntax expected\n");
        printf("    $ test_init.exe <INPUT FILE> <OUTPUT FILE> <SOLVER> [<ELECTRIC FIELD FILE>]\n");
        return 1;
    }
    
    int n_ipd, pol_atoms, solver;
    double *_ipd;
    double ***ipd, *electric_field;
    double **external_ef;
    int32_t *polar_mm;

    OMMP_SYSTEM_PRT my_system = ommp_init_mmp(argv[1]);
    // ommp_set_verbose(OMMP_VERBOSE_DEBUG);

    n_ipd = ommp_get_n_ipd(my_system);
    pol_atoms = ommp_get_pol_atoms(my_system);

    if(strcmp(argv[3], "inversion") == 0){
        printf("Solving with matrix inversion\n");
        solver = OMMP_SOLVER_INVERSION;
    }
    else if(strcmp(argv[3], "cg") == 0){
        printf("Solving with conjugate gradients\n");
        solver = OMMP_SOLVER_CG;
    }
    else if(strcmp(argv[3], "diis") == 0){
        printf("Solving with Jacobi-DIIS\n");
        solver = OMMP_SOLVER_DIIS;
    }
    else{
        printf("Unrecognized solver (\"%s\"). Exiting.\n", argv[3]);
        return 1;
    }
    
    electric_field = (double *) malloc(sizeof(double) * 3 * pol_atoms);
    polar_mm = (int32_t *) ommp_get_polar_mm(my_system);
   
    if(argc == 5)
        external_ef = read_ef(argv[4]);

    for(int j = 0; j < pol_atoms; j++)
        for(int k = 0; k < 3; k++)
            if(argc == 5)
                electric_field[j*3+k] = external_ef[polar_mm[j]-1][k];
            else
                electric_field[j*3+k] = 0.0;

    ommp_set_external_field(my_system, electric_field, solver);

    // Get induced point dipoles
    _ipd = (double *) ommp_get_ipd(my_system);
    ipd = (double ***) malloc(sizeof(double **) * n_ipd );

    for(int i = 0; i < n_ipd; i++){
        ipd[i] = (double **) malloc(sizeof(double *) * pol_atoms);

        for(int j = 0; j < pol_atoms; j++){
            ipd[i][j] = &(_ipd[i*pol_atoms*3 + j*3]);
        }
    }

    FILE *fp = fopen(argv[2], "w+");

    for(int k = 0; k < n_ipd; k++){
        // fprintf(fp, "IPD SET %d\n", k+1);
        
        for(int i = 0; i < pol_atoms; i++){
            for(int j = 0; j < 3; j++)
                fprintf(fp, "%20.12e", ipd[k][i][j]);
            fprintf(fp, "\n");
        }
    }
    
    fclose(fp);
    free(electric_field);
    
    for(int i = 0; i < n_ipd; i++)
        free(ipd[i]);
    free(ipd);
    
    if(argc == 5)
        free(external_ef);
    ommp_terminate(my_system);
    
    return 0;
}

File: src/test_programs/F03/test_ipd.f90
program test_ipd
    use iso_c_binding, only: c_char
    use ommp_interface

    implicit none

    integer :: argc, i, j, k, solver
    character(kind=c_char, len=120), dimension(4) :: argv
    real(ommp_real), allocatable :: ef(:,:), ef_pol(:,:)
    type(ommp_system), pointer :: my_system

    argc = command_argument_count()
    if(argc /= 3 .and. argc /= 4 ) then
        write(*, *) "Syntax expected"
        write(*, *) "    $ test_init.exe <INPUT FILE> <OUTPUT FILE> <SOLVER>&
                    & [<ELECTRIC FIELD FILE>]"
        stop 1
    end if

    do i=1, argc
        call get_command_argument(i, argv(i))
    end do

    call ommp_set_verbose(OMMP_VERBOSE_DEBUG)
    call ommp_init_mmp(my_system, argv(1))

    select case(argv(3))
        case("cg")
            solver = OMMP_SOLVER_CG
        case("inversion")
            solver = OMMP_SOLVER_INVERSION
        case("diis")
            solver = OMMP_SOLVER_DIIS
        case default
            write(*, "('Unrecognized solver (',A,'). Exiting.')") argv(3)
            stop 1
    end select

    allocate(ef(3, my_system%top%mm_atoms))
    allocate(ef_pol(3, my_system%eel%pol_atoms))
    
    if(argc == 4) then
        open(unit=101, file=argv(4)) 
        do i=1, my_system%top%mm_atoms
            read(101, *) ef(:, i)
        end do
        close(101)
    else
        ef = 0.0
    end if

    do i=1, my_system%eel%pol_atoms
        ef_pol(:,i) = ef(:, my_system%eel%polar_mm(i))
    end do
    
    deallocate(ef)
    call ommp_set_external_field(my_system, ef_pol, solver)

    ! Write ipd
    open(unit=101, file=argv(2))
    do k = 1, my_system%eel%n_ipd
        do j = 1, my_system%eel%pol_atoms
            write(101, "(3F20.12)") my_system%eel%ipd(:, j, k)
        end do
    end do
    close(101)

    deallocate(ef_pol)
    call ommp_terminate(my_system)

end program test_ipd
